apps/server/lib/lexical/server/transport/std_io.ex

```diff
@@ -76,29 +83,62 @@ defmodule Lexical.Server.Transport.StdIO do
     end
   end
 
-  defp parse_headers(headers) do
-    Enum.map(headers, &parse_header/1)
-  end
+  defp content_length(headers) do
+    case List.keyfind(headers, "content-length", 0) do
+      {_, len_str} ->
+        {:ok, String.to_integer(len_str)}
 
-  defp header_value(headers, header_name, converter) do
-    case List.keyfind(headers, header_name, 0) do
-      nil -> :error
-      {_, value} -> {:ok, converter.(value)}
+      nil ->
+        {:error, :no_content_length}
     end
   end
 
   defp read_body(device, byte_count) do
     case IO.binread(device, byte_count) do
-      data when is_binary(data) or is_list(data) ->
+      data when is_binary(data) and byte_size(data) === byte_count ->
         {:ok, data}
 
-      other ->
-        other
+      data when is_binary(data) ->
+        read_more(device, byte_count - byte_size(data), data)
+
+      data when is_list(data) ->
+        if Enum.count(data) === byte_count do
+          {:ok, data}
+        else
+          remaining_bytes = byte_count - byte_size(data)
+          read_more(device, remaining_bytes, data)
+        end
+
+      :eof ->
+        {:error, :eof}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  # give editor an extra ms if there's not enough bytes.
+  defp read_more(device, byte_count, data) do
+    :timer.sleep(1)
+
+    case IO.binread(device, byte_count) do
+      rest when is_binary(rest) or is_list(rest) ->
+        [data | rest]
+
+      :eof ->
+        {:error, :eof}
+
+      error ->
+        error
     end
   end
 
+  defp parse_headers(headers) do
+    Enum.map(headers, &parse_header/1)
+  end
+
   defp parse_header(line) do
-    [name, value] = String.split(line, ":")
+    [name, value] = String.split(line, ":", parts: 2)
```

Comment:
i'm r- on this. Headers should have two parts.  If there are more, something is seriously wrong.

Reply:
removed